### ISSUE TITLE
ML 2 Week 1 Project Feedback

### ISSUE THREAD

**Issue by bonareri** [Njeri-Kiarie/MLII_dimensionality_reduction#1]
<html>
<body>
<html><head></head><body>
<h2>ML 2 Week 1 Project Feedback</h2><h3>Score: 94/125 — 75.2%</h3><h3>What you did well</h3><ul><li><p>You followed a clear workflow covering data preparation, feature engineering, encoding, scaling, PCA, modelling, and evaluation.</p></li><li><p>Your investigation of missing values was thorough. You did not immediately assume that missing values meant the absence of a feature; you investigated their distribution across several variables before deciding how to handle them.</p></li><li><p>You correctly used placeholder categories such as <code inline="">Unknown</code> for missing <code inline="">Variant</code> values and <code inline="">None</code> for missing <code inline="">Accessories</code> values.</p></li><li><p>You created useful features including <strong>Car Age, Serviced km, Has Full Service History, Has Service Record, and Number of Accessories</strong>.</p></li><li><p>Importantly, although you created <strong>Price per Mile</strong> as required by the project, you correctly excluded it from the modelling features because it is derived directly from the target variable, <code inline="">Price</code>. This was a good modelling decision.</p></li><li><p>You performed the train-test split before encoding and scaling, which is the correct direction for avoiding data leakage.</p></li><li><p>You correctly identified that the Random Forest model was overfitting, with a training R² of approximately 0.857 compared with a negative testing R².</p></li><li><p>Your conclusion acknowledges that dimensionality reduction does not automatically improve predictive performance, which is an important ML insight.</p></li></ul><h3>Areas to Improve</h3><h4>1. Be accurate about the correlation analysis</h4><p>You state that:</p><blockquote><p>"Highly correlated or redundant features, such as Make Year and Car Age, were identified and one was removed."</p></blockquote><p>However, <code inline="">Make Year</code> had already been removed before the correlation analysis was performed. Therefore, the correlation matrix could not have been used to identify <code inline="">Make Year</code> and <code inline="">Car Age</code> as highly correlated at that stage.</p><p>You should distinguish between:</p><ul><li><p>features removed during feature engineering, and</p></li><li><p>features actually identified as redundant through the correlation analysis.</p></li></ul><p>Also, if no features were removed based on the correlation threshold, state that clearly.</p><h4>2. Your correlation analysis needs a defined threshold</h4><p>You present the correlation matrix and discuss multicollinearity, but you do not clearly show a numerical threshold used to determine which features are highly correlated.</p><p>Instead of relying only on the heatmap, define a threshold and identify the feature pairs that meet it.</p><p>For example, you could investigate correlations above an absolute value such as <code inline="">0.85</code> and explicitly state whether any features meet that threshold.</p><h4>3. Be careful with the <code inline="">numerical_cols</code> definition</h4><p>You define:</p><pre><code class="language-python">numerical_cols = [
    'Mileage',
    'Car Age',
    'Price per Mile'
]</code></pre><p>But <code inline="">Price per Mile</code> has already been excluded from <code inline="">X</code>, so it is not actually present in the modelling dataset at this point.</p><p>More importantly, you subsequently scale <strong>all columns in <code inline="">X_train</code></strong>, rather than using <code inline="">numerical_cols</code>.</p><p>Make sure your code reflects the preprocessing strategy you are describing. Avoid defining a list of numerical features that contains a variable that is no longer part of the model input.</p><h4>4. Think carefully about what should be standardized before PCA</h4><p>You apply <code inline="">StandardScaler</code> to the entire encoded feature matrix, including binary and one-hot encoded variables.</p><p>Scaling before PCA is important because PCA is sensitive to feature scale. However, one-hot/binary variables require some consideration because they are categorical representations rather than continuous measurements.</p><p>Explain your preprocessing choice and why you decided to apply the scaler to the full feature matrix before PCA.</p><p>A <code inline="">ColumnTransformer</code> would also make the preprocessing steps clearer and more systematic.</p><h4>5. Your PCA reduction is relatively modest</h4><p>You reduced the feature space from <strong>41 features to 34 components</strong>, retaining approximately <strong>95.59% of the variance</strong>.</p><p>This is valid, but it only reduces the feature space by 7 features.</p><p>Discuss whether retaining 95% of the variance is the most appropriate objective for this predictive task. You could also compare different variance thresholds and investigate whether a smaller number of components provides a useful trade-off between information retention and dimensionality reduction.</p><h4>6. Your comparison of PCA vs non-PCA is incomplete</h4><p>You compare Linear Regression before and after PCA, which is useful.</p><p>However, your Random Forest and Gradient Boosting models are evaluated only on the PCA-transformed data.</p><p>Therefore, you cannot conclude how PCA affected the performance of Random Forest or Gradient Boosting.</p><h4>7. Interpret the negative R² correctly</h4><p>You correctly recognize that the negative test R² values indicate poor predictive performance.</p><p>However, remember that a negative R² means that the model performs worse than a simple baseline that predicts the mean target value for every observation.</p><p>Your Linear Regression model had the least negative test R² (<code inline="">-0.018</code>), making it the strongest of the three <strong>on the test set</strong>, but its performance is still poor.</p><p>Avoid presenting Linear Regression as a good predictive model simply because it performed best relative to the other models.</p><h4>8. Investigate why the models generalize so poorly</h4><p>The most important modelling finding in your project is that <strong>all three models have negative test R² values</strong>.</p><p>Rather than stopping at identifying poor generalization, investigate possible reasons.</p><p>For example:</p><ul><li><p>Are the available features sufficiently informative for predicting price?</p></li><li><p>Is the target distribution highly variable?</p></li><li><p>Are there important variables missing from the dataset?</p></li><li><p>Does PCA remove useful predictive structure?</p></li><li><p>Are there nonlinear relationships that the current models are failing to capture?</p></li><li><p>Are there outliers affecting the evaluation?</p></li></ul><p>This would make the analysis much more insightful.</p><h4>9. Improve the service-history feature engineering</h4><p>Your extraction of <code inline="">Serviced km</code> is a good idea, but assigning <code inline="">0</code> when there is no mileage information means that several different situations can end up with the same value.</p><p>For example:</p><pre><code class="language-text">No Service Record → 0
Full Service History → 0
Actual extracted service mileage → numeric value</code></pre><p>The <code inline="">Has Service Record</code> and <code inline="">Has Full Service History</code> variables help distinguish some of these cases, but explain this design more clearly and ensure that the meaning of <code inline="">0</code> is not misunderstood.</p><h4>10. Strengthen the model evaluation</h4><p>You have reported MAE, RMSE, and R², which is good.</p><p>Go further by including:</p><ul><li><p>Actual vs predicted prices</p></li><li><p>Residual plots</p></li><li><p>Distribution of residuals</p></li><li><p>Identification of observations with very large errors</p></li></ul><p>This would help you understand why the models are performing poorly rather than only reporting the metrics.</p><h3>Overall Feedback</h3><p>You have demonstrated a good understanding of the overall machine learning workflow, particularly in your handling of missing values, feature engineering, train-test splitting, target leakage, PCA, and model evaluation.</p><p>One of the strongest parts of your work is recognizing that <strong>Price per Mile should not be used as a predictor when Price is the target</strong>. You also correctly performed the train-test split before the main preprocessing steps.</p><p>The main areas to improve are the <strong>accuracy of your methodological explanations and the depth of your analysis</strong>. In particular, make sure that your written interpretation matches what your code actually did. Your correlation analysis, PCA comparison, and explanation of the model performance need more careful interpretation.</p><p>Your models did not perform well on the test set, but that is not necessarily a failure of the project. What matters is that you investigate <strong>why</strong> they performed poorly and communicate those findings accurately.</p><p>Good work overall. Focus on making your methodological reasoning more precise and strengthening the analysis of why the models are struggling to generalize.</p></body></html>
</body>
</html>
